crates/esthri/src/ops/sync.rs

```diff
@@ -169,77 +170,86 @@ fn process_globs<'a, P: AsRef<Path> + 'a>(path: P, filters: &[GlobFilter]) -> Op
+    tokio::task::spawn_blocking(move || async move {
+        for entry in walk_dir {
+            match entry {
+                Ok(entry) => {
+                    if entry
+                        .file_name()
+                        .to_string_lossy()
+                        .contains(TEMP_FILE_PREFIX)
+                    {
+                        continue;
+                    }
+                    if entry.path_is_symlink() {
+                        warn!("symlinks are ignored");
+                        continue;
+                    }
+                    debug!("local path={}", entry.path().display());
+                    if process_globs(entry.path(), &filters).is_some() {
+                        tx.send(Ok(entry.into_path()))
+                            .expect("failed to send file path");
+                    }
+                }
+                Err(e) => {
+                    tx.send(Err(e.into())).expect("failed to send error");
+                    break;
+                }
             }
         }
-    }
+    });
+    UnboundedReceiverStream::new(rx)
 }
 
 #[derive(Debug, Clone, Copy)]
-enum SyncCompressionDirection {
-    None,
+enum SyncCmd {
     Up,
+    UpCompressed,
     Down,
+
+    DownCompressed,
 }
 
 fn translate_paths<StreamT>(
     input_stream: StreamT,
-    compressed: SyncCompressionDirection,
+    sync_cmd: SyncCmd,
 ) -> impl Stream<Item = impl Future<Output = MapPathResult>>
 where
-    StreamT: Stream<Item = Result<(String, Option<ListingMetadata>)>>,
+    StreamT: Stream<Item = Result<(PathBuf, Option<ListingMetadata>)>>,
 {
-    use std::str::FromStr;
     input_stream.map(move |params| async move {
-        let (path, metadata) = params?;
-        let source_path = PathBuf::from_str(&path)?;
-        let file_path: Box<dyn AsRef<Path>> = Box::new(path);
+        let (source_path, metadata) = params?;
+        let file_path: Box<dyn AsRef<Path>> = Box::new(source_path.clone());
 
-        let (file_path, source_path, local_etag) = match compressed {
-            SyncCompressionDirection::None => {
-                let local_etag = compute_etag(&source_path).await;
+        let (file_path, source_path, local_etag) = match sync_cmd {
+            SyncCmd::Up => {
+                let local_etag = if metadata.is_some() {
+                    compute_etag(&source_path).await.map(Option::Some)
+                } else {
+                    Ok(None)
+                };
                 (file_path, source_path, local_etag)
             }
-            SyncCompressionDirection::Up => {
+            SyncCmd::UpCompressed => {
                 // If we're syncing up with compression, then everything
                 // should be compressed
                 let (tmp, _) = compress_to_tempfile(&source_path).await?;
-                let local_etag = compute_etag(tmp.path()).await;
+                let local_etag = if metadata.is_some() {
+                    compute_etag(&source_path).await.map(Option::Some)
```

Comment:
```suggestion
                    compute_etag(tmp.path()).await.map(Option::Some)
```

I think we still need to be computing the etag of the compressed file here.

(Probably also reveals that we might need better `sync` test coverage - although admittedly it's tricky to test what a sync command actually has done atm)

Reply:
Ahh good catch. Yeah we could definitely improve on the testing situation